instrumentation/apache-httpclient/apache-httpclient-4.3/library/src/main/java/io/opentelemetry/instrumentation/apachehttpclient/v4_3/TracingProtocolExec.java

```diff
@@ -0,0 +1,172 @@
+      // Request already had a context so a redirect. Don't create a new span just inject and
+      // execute.
+      propagators.getTextMapPropagator().inject(context, request, HttpHeaderSetter.INSTANCE);
+      return execute(route, request, instrumenterRequest, httpContext, httpExecutionAware, context);
+    }
+
+    HttpHost host = null;
+    if (route.getTargetHost() != null) {
+      host = route.getTargetHost();
+    } else if (httpContext.getTargetHost() != null) {
+      host = httpContext.getTargetHost();
+    }
+    if (host != null) {
+      if ((host.getSchemeName().equals("https") && host.getPort() == 443)
+          || (host.getSchemeName().equals("http") && host.getPort() == 80)) {
+        // port seems to be added to the host by route planning for standard ports even if not
+        // specified in the URL. There doesn't seem to be a way to differentiate between explicit
+        // and implicit port, but ignore in both cases to match the more common case.
+        host = new HttpHost(host.getHostName(), -1, host.getSchemeName());
+      }
+    }
+    ApacheHttpClientRequest instrumenterRequest = new ApacheHttpClientRequest(host, request);
+
+    Context parentContext = Context.current();
+    if (!instrumenter.shouldStart(parentContext, instrumenterRequest)) {
+      return exec.execute(route, request, httpContext, httpExecutionAware);
+    }
+
+    context = instrumenter.start(parentContext, instrumenterRequest);
+    httpContext.setAttribute(REQUEST_CONTEXT_ATTRIBUTE_ID, context);
+    httpContext.setAttribute(REQUEST_WRAPPER_ATTRIBUTE_ID, instrumenterRequest);
+    httpContext.setAttribute(REDIRECT_COUNT_ATTRIBUTE_ID, 0);
+
+    propagators.getTextMapPropagator().inject(context, request, HttpHeaderSetter.INSTANCE);
+
+    return execute(route, request, instrumenterRequest, httpContext, httpExecutionAware, context);
+  }
+
+  private CloseableHttpResponse execute(
+      HttpRoute route,
+      HttpRequestWrapper request,
+      ApacheHttpClientRequest instrumenterRequest,
+      HttpClientContext httpContext,
+      HttpExecutionAware httpExecutionAware,
+      Context context)
+      throws IOException, HttpException {
+    CloseableHttpResponse response = null;
+    Throwable error = null;
+    try (Scope ignored = context.makeCurrent()) {
+      response = exec.execute(route, request, httpContext, httpExecutionAware);
+      return response;
+    } catch (Throwable e) {
+      error = e;
+      throw e;
+    } finally {
+      if (!pendingRedirect(context, httpContext, request, instrumenterRequest, response)) {
+        instrumenter.end(context, instrumenterRequest, response, error);
+      }
+    }
+  }
+
+  private boolean pendingRedirect(
+      Context context,
+      HttpClientContext httpContext,
+      HttpRequestWrapper request,
+      ApacheHttpClientRequest instrumenterRequest,
+      @Nullable CloseableHttpResponse response)
+      throws ProtocolException {
```

Comment:
This method can throw and is called in a `finally` block - would it make sense to catch whatever it throws and `addSuppressed()` to `error`? Or is it a case when you have to declare `throws` but nothing should ever fail here?

Reply:
Thanks for noticing! Tightened up the below catch to prevent it from leaking, which would be bad for many reasons not just being in a finally (well technically being in the finally would save the app from crashing maybe :)